File: 백준/Gold/3079. 입국심사/입국심사.py
N, M = map(int , input().split())
T = []

start = 0
end = 10**18 # 초기화

for _ in range(N):
    a = int(input())
    T.append(a)

# print(T)
def check(t): 
    # a의 시간에 모든 사람이 통과할 수 있는지 확인
    # 어떻게 확인하지?
    result = 0
    for k in T:
        temp = (t // k)
        result += temp
    if (result >= M):
        return True
    else:
        return False
        


while start < end:
    mid = (start + end) // 2
    if (check(mid)): # 통과가능이면
        end = mid
    else:
        start = mid + 1
print( end)

File: 프로그래머스/3/43238. 입국심사/입국심사.cpp
#include <string>
#include <vector>
#include <algorithm>
// #include <cmath>
#include <math.h>
#include <iostream>

using namespace std;

bool check(long long mid, vector<int> times, int n ) {
    long long temp = 0;
    for (auto& t : times) {
        // cout << t << " "; 
        // cout << mid << " "; 
        temp += mid / t;
    }
    
    // cout << temp << endl; 
    
    if (temp >= n) return true;
    else return false;
}

long long solution(int n, vector<int> times) {
    long long answer = 0;
    
    // 가능한 최솟값을 return한다. 
    long long end = pow(10, 18);
    long long start = 0;
    
    while (start < end) {
        long long mid = (start + end) /2 ;
        
        if (check(mid, times, n) == true) {
            end = mid ;
        } else {
            start = mid + 1;
        }
        // break; 
    }
    
    answer = start; 
    return answer;
}